### PR TITLE
fix: Restructure WidgetNameCanvas

### LINKED ISSUE #28201: [Task]: Fix Anvil Widget Name Component issues
### Is there an existing issue for this?

- [X] I have searched the existing issues

### SubTasks

- Multiple widgets don't  show widget names when selected
- Widget name changes offset when scrolling
- Clicking on the widget name doesn't select the widget

### PR DESCRIPTION
**Pull request by riodeuno**
#### PR fixes following issue(s)
Fixes #28201

#### Type of change
- New feature (non-breaking change which adds functionality) - doesn't effect user perception
#### How Has This Been Tested?
- [ ] Manual
- [ ] JUnit
- [ ] Jest
- [ ] Cypress
#### Dev activity
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

### CODE REVIEW

File: app/client/src/layoutSystems/common/WidgetNamesCanvas/widgetNameRenderUtils.ts [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+  WidgetNameData,
+  WidgetNamePositionType,
+} from "./WidgetNameTypes";
+import { SelectionRequestType } from "sagas/WidgetSelectUtils";
+import { getWidgetNameComponent } from "./utils";
+import type { KonvaEventListener } from "konva/lib/Node";
+import type { Group } from "konva/lib/Group";
+
+export function getMainContainerAnvilCanvasDOMElement() {
+  return document.getElementById(`anvil-canvas-0`) as HTMLDivElement | null;
```
Comment (riodeuno) [inline app/client/src/layoutSystems/common/WidgetNamesCanvas/widgetNameRenderUtils.ts:17]: @riodeuno The id getter function is available in `app/client/src/layoutSystems/anvil/canvas/utils.ts` 

`getAnvilCanvasId(MAIN_CONTAINER_WIDGET_ID)`

File: app/client/package.json [hunk 2]
```diff
@@ -85,6 +85,7 @@
     "astring": "^1.7.5",
     "aws-sdk": "^2.1443.0",
     "axios": "^0.27.2",
+    "canvas": "^2.11.2",
```
Comment (riodeuno) [inline app/client/package.json:88]: Check if this is only a dev dependency.

File: app/client/src/layoutSystems/common/WidgetNamesCanvas/index.tsx [hunk 3]
```diff
@@ -99,266 +109,86 @@ const OverlayCanvasContainer = (props: {
...
-
-    return { isMouseOver: false };
-  };
+  // Reset the canvas if no widgets are focused or selected
+  // Update the widget name positions if there are widgets focused or selected
+  // and they've changed.
 
-  //Used when the position of selected or focused widget changes
+  // Note: If the selector for `selectWidgetNameData` reference changes
+  // Then this will run on every render. We should be careful about this.
```
Comment (riodeuno) [inline app/client/src/layoutSystems/common/WidgetNamesCanvas/index.tsx:176]: Verify the performance and see how many times it is called.

File: app/client/src/pages/Editor/WidgetsEditor/MainContainerWrapper.tsx [hunk 4]
```diff
@@ -250,6 +254,12 @@ function MainContainerWrapper(props: MainCanvasWrapperProps) {
           </div>
         )}
         {node}
+        {enableOverlayCanvas && (
+          <OverlayCanvasContainer
+            canvasWidth={canvasWidth}
+            containerRef={props.parentRef}
```
Comment (riodeuno) [inline app/client/src/pages/Editor/WidgetsEditor/MainContainerWrapper.tsx:260]: Verify if containerRef is necessary.

**Review by prsidhu — APPROVED** [review]

**Review by marks0351 — APPROVED** [review]

**Pull request merged**